src/Features/Core/Portable/InlineMethod/AbstractInlineMethodRefactoringProvider.cs

```diff
@@ -0,0 +1,373 @@
+            SyntaxNode callerDeclarationNode,
+            ISymbol callerSymbol,
+            bool containsAwaitExpression,
+            InlineMethodContext inlineMethodContext,
+            bool removeCalleeDeclarationNode,
+            CancellationToken cancellationToken)
+        {
+            var solution = document.Project.Solution;
+            var solutionEditor = new SolutionEditor(solution);
+            var documentEditor = await solutionEditor.GetDocumentEditorAsync(document.Id, cancellationToken).ConfigureAwait(false);
+            var statementContainsCalleeInvocationExpression = inlineMethodContext.StatementContainingCallee;
+            foreach (var statement in inlineMethodContext.StatementsToInsertBeforeCallee)
+            {
+                documentEditor.InsertBefore(
+                    statementContainsCalleeInvocationExpression,
+                    // Make sure the statement is aligned with the existing statement
+                    statement.WithTriviaFrom(statementContainsCalleeInvocationExpression));
+            }
+
+            var syntaxNodeToReplace = inlineMethodContext.SyntaxNodeToReplace;
+            var inlineSyntaxNode = inlineMethodContext.InlineSyntaxNode;
+            documentEditor.ReplaceNode(syntaxNodeToReplace, inlineSyntaxNode);
+
+            // If the inline content has 'await' expression, then make sure the caller is converted to 'async' method
+            if (containsAwaitExpression)
+            {
+                // If the inlineExpresion contains await Expression, check has been done to make sure
+                // the caller is Method
+                var callerMethodSymbol = (IMethodSymbol)callerSymbol;
+                if (!callerMethodSymbol.IsAsync)
+                {
+                    var declarationModifiers = DeclarationModifiers.From(callerSymbol).WithAsync(true);
+                    documentEditor.SetModifiers(callerDeclarationNode, declarationModifiers);
+                }
+            }
+
+            if (removeCalleeDeclarationNode)
+            {
+                var documentId = solution.GetDocumentId(calleeMethodNode.SyntaxTree);
+                if (documentId != null)
+                {
+                    var editor = await solutionEditor.GetDocumentEditorAsync(documentId, cancellationToken).ConfigureAwait(false);
+                    editor.RemoveNode(calleeMethodNode);
+                }
+            }
+
+            return solutionEditor.GetChangedSolution();
+        }
+
+        private class MySolutionChangeAction : CodeAction.SolutionChangeAction
+        {
+            public MySolutionChangeAction(
+                string title,
+                 Func<CancellationToken, Task<Solution>> createChangedSolution,
+                string? equivalenceKey = null) : base(title, createChangedSolution, equivalenceKey)
+            {
+            }
+        }
+
+        private class MyNestedCodeAction : CodeAction.CodeActionWithNestedActions
```

Comment:
you don't need this.  you can just use a normal code action for the top-evel item :)